src/MultiTokenPermit.sol

```diff
@@ -0,0 +1,237 @@
+     * @param token Token contract address
+     * @param spender Address to approve
+     * @param tokenId Token ID (0 for ERC20, specific ID for NFT, type(uint256).max for collection wildcard)
+     * @param amount Amount to approve (ignored for ERC721, used for ERC20/ERC1155)
+     * @param expiration Timestamp when approval expires (0 for no expiration)
+     */
+    function approve(
+        address token,
+        address spender,
+        uint256 tokenId,
+        uint160 amount,
+        uint48 expiration
+    ) external override {
+        address tokenKey;
+
+        if (tokenId == 0 || tokenId == type(uint256).max) {
+            // ERC20 or collection-wide wildcard
+            tokenKey = token;
+        } else {
+            // Specific token ID - encode it as an address
+            tokenKey = address(uint160(uint256(keccak256(abi.encodePacked(token, tokenId)))));
+        }
+
+        // Update the allowance
+        allowances[msg.sender][tokenKey][spender] =
+            Allowance({ amount: amount, expiration: expiration, timestamp: uint48(block.timestamp) });
+
+        // Emit standard approval event from IPermit interface
+        emit Approval(msg.sender, tokenKey, spender, amount, expiration);
+    }
+
+    /**
+     * @notice Execute approved ERC721 token transfer
+     * @dev Uses the dual-allowance system: tries per-token allowance first, falls back to collection-wide
+     * @param from Token owner address
+     * @param to Transfer recipient address
+     * @param token ERC721 contract address
+     * @param tokenId The unique NFT token ID
+     */
+    function transferFrom(address from, address to, address token, uint256 tokenId) public override {
+        // Create unique encoded identifier for this specific token ID
+        // Uses keccak256(token || tokenId) truncated to address for deterministic mapping
+        address encodedId = address(uint160(uint256(keccak256(abi.encodePacked(token, tokenId)))));
+
+        // First, try to update allowance for the specific token ID
+        (, bytes memory revertDataPerId) = _updateAllowance(from, encodedId, msg.sender, 1);
+
+        if (revertDataPerId.length > 0) {
+            // Fallback: try collection-wide allowance if per-token-id allowance fails
+            (, bytes memory revertDataWildcard) = _updateAllowance(from, token, msg.sender, 1);
+            if (revertDataWildcard.length > 0) {
+                // Priority error handling: show collection-wide error for insufficient allowance,
+                // otherwise show the more specific per-token error
+                bytes4 perIdSelector = bytes4(revertDataPerId);
+                if (perIdSelector == InsufficientAllowance.selector) {
+                    _revert(revertDataWildcard);
+                } else {
+                    _revert(revertDataPerId);
+                }
+            }
+        }
+        IERC721(token).safeTransferFrom(from, to, tokenId);
+    }
```

Comment:
I am confused here. Why not only try the wildcard when the token Id is `type(uint256).max`, which is more consistent with the `approve` function.